src/common/yaml/platform/yaml-cpp/yaml.cpp

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <common/yaml.hpp>
+
+#include <fstream>
+#include <map>
+#include <string>
+
+#include <yaml-cpp/yaml.h>
+
+#include <common/io.hpp>
+
+using namespace std;
+
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace io = mender::common::io;
+
+namespace mender {
+namespace common {
+namespace yaml {
+
+ExpectedYaml LoadFromFile(string file_path) {
+	ifstream f;
+	errno = 0;
+	f.open(file_path);
+	if (not f) {
+		int io_errno = errno;
+		auto err = error::Error(
+			std::generic_category().default_error_condition(io_errno),
+			"Failed to open '" + file_path + "': " + strerror(io_errno));
+		return expected::unexpected(err);
+	}
+
+	try {
+		YAML::Node loaded_yaml = YAML::LoadFile(file_path);
+		return loaded_yaml;
+	} catch (YAML::Exception &e) {
+		return expected::unexpected(
+			MakeError(ParseError, "Failed to parse '" + file_path + "'" + e.what()));
+	}
+}
+
+ExpectedYaml Load(string yaml_str) {
+	try {
+		YAML::Node loaded_yaml = YAML::Load(yaml_str);
+		return loaded_yaml;
+	} catch (YAML::Exception &e) {
+		return expected::unexpected(
+			MakeError(ParseError, "Failed to parse '" + yaml_str + "'" + e.what()));
+	}
+}
+
+ExpectedYaml Load(istream &str) {
+	try {
+		YAML::Node loaded_yaml = YAML::Load(str);
+		return loaded_yaml;
+	} catch (YAML::Exception &e) {
+		return expected::unexpected(
+			MakeError(ParseError, "Failed to parse YAML from stream: " + string(e.what())));
+	}
+}
+
+ExpectedYaml Load(io::Reader &reader) {
+	auto str_ptr = reader.GetStream();
+	return Load(*str_ptr);
+}
+
+bool Yaml::IsObject() const {
+	return n_yaml.IsMap();
+}
+bool Yaml::IsArray() const {
+	return n_yaml.IsSequence();
+}
+
+template <typename T>
+bool Is(const YAML::Node &n) {
+	if (not n.IsScalar()) {
```

Comment:
Isn't `IsScalar` false for arrays? And will it then work to call `Is<vector<string>>(n)`? Or do we simply not care about complex types in this function (should maybe be commented then)?

Reply:
As is now, it's just for base types, and no collections. I'll leave a comment.

Reply:
Actually, I'll just pass everything on. Then it will work for composite types also. Thx for picking up on this